onnxruntime/core/providers/cuda/reduction/reduction_ops.cc

```diff
@@ -455,27 +455,52 @@ Status ReduceComputeCore(CUDAExecutionProvider& cuda_ep, const Tensor& input, Pr
+      const CudaT* input_data = reinterpret_cast<const CudaT*>(input.template Data<T>());
+      if (calculate_sqt) {
+        input_data_buffer = cuda_ep.GetScratchBuffer<T>(input_count);
+        input_data = reinterpret_cast<CudaT*>(input_data_buffer.get());
+        fast_divmod tmp_div;
+        Impl_Mul<CudaT>(stream, static_cast<int32_t>(SimpleBroadcast::NoBroadcast), nullptr,
+                        reinterpret_cast<const CudaT*>(input.template Data<T>()), nullptr,
+                        reinterpret_cast<const CudaT*>(input.template Data<T>()), nullptr, tmp_div, tmp_div,
+                        reinterpret_cast<CudaT*>(input_data_buffer.get()), input_count);
+        input_data = reinterpret_cast<const CudaT*>(input_data_buffer.get());
       }
-      case ApplicableMatrixReduction::Columns: {
-        const auto buffer_size_bytes = compute_reduce_matrix_columns_buffer_size<CudaT>(m, n);
-        auto buffer = cuda_ep.GetScratchBuffer<void>(buffer_size_bytes);
-        return reduce_matrix_columns(
-            stream,
-            reinterpret_cast<const CudaT*>(input.template Data<T>()),
-            reinterpret_cast<CudaT*>(output.template MutableData<T>()),
-            m, n, buffer.get(), buffer_size_bytes);
+
+      switch (applicable_matrix_reduction) {
+        case ApplicableMatrixReduction::Rows: {
+          ORT_RETURN_IF_ERROR(reduce_matrix_rows(
+              stream, input_data, reinterpret_cast<CudaT*>(output.template MutableData<T>()), m, n));
+        } break;
+        case ApplicableMatrixReduction::Columns: {
+          const auto buffer_size_bytes = compute_reduce_matrix_columns_buffer_size<CudaT>(m, n);
+          auto buffer = cuda_ep.GetScratchBuffer<void>(buffer_size_bytes);
+          ORT_RETURN_IF_ERROR(reduce_matrix_columns(stream, input_data,
+                                                    reinterpret_cast<CudaT*>(output.template MutableData<T>()), m, n,
+                                                    buffer.get(), buffer_size_bytes));
+        } break;
+        default: {
+          ORT_ENFORCE(false, "Invild matrix reduction type.");
+        }
       }
-      default:
-        break;
+
+      if (calculate_log) {
+        Impl_Log<CudaT>(stream, reinterpret_cast<const CudaT*>(output.template Data<T>()),
+                        reinterpret_cast<CudaT*>(output.template MutableData<T>()), output_count);
+      } else if (cudnn_reduce_op == CUDNN_REDUCE_TENSOR_AVG) {
+        float denominator_float = applicable_matrix_reduction == ApplicableMatrixReduction::Rows
+                                      ? static_cast<float>(m)
+                                      : static_cast<float>(n);
+        CudaT denominator = ToCudaType<T>::FromFloat(denominator_float);
+        UnaryDiv(stream, reinterpret_cast<const CudaT*>(output.template Data<T>()),
+                 reinterpret_cast<CudaT*>(output.template MutableData<T>()), denominator, output_count);
```

Comment:
For reduce to one scalar case, can the reduce_mean defined here https://github.com/microsoft/onnxruntime/blob/0bb56a18cf183343fe2aa62dbcf3130fc3308c76/onnxruntime/core/providers/cuda/reduction/reduction_functions.cu#L338 be used?

Reply:
I did a quick test on the scalar case using 3 ways to calculate mean on [4096, 4096] tensor:

cuDNN's full_reduce_tensor_kernel: ~88us
reduce_metrix_rows_kernel+UnaryDiv: ~89us+2us=91us
reduce_mean (which calls reduce_matrix_columns_kernel with DivideResultBySize=true): ~95us

Acutally cuDNN handles such case well. reduce_metrix_rows_kernel is indeed faster than reduce_matrix_columns_kernel. But overall I think the perf diff is small. So I am guessing current way of reduce_metrix_rows_kernel+UnaryDiv should be fine, and it makes the code simpler.

Reply:
OK, let's get this PR merged first since release 1.9 is already branched. We can optimize them further later if needs.